server/cluster/unsafe_recovery_controller.go

```diff
@@ -657,27 +693,39 @@ func (r *regionItem) IsEpochStale(other *regionItem) bool {
 	return re.GetVersion() < oe.GetVersion() || re.GetConfVer() < oe.GetConfVer()
 }
 
-func (r *regionItem) IsRaftStale(origin *regionItem) bool {
-	rs := r.report.GetRaftState()
-	os := origin.report.GetRaftState()
-	if rs.GetHardState().GetTerm() < os.GetHardState().GetTerm() {
-		return true
-	} else if rs.GetHardState().GetTerm() == os.GetHardState().GetTerm() {
-		if rs.GetLastIndex() < os.GetLastIndex() {
-			return true
-		} else if rs.GetLastIndex() == os.GetLastIndex() {
-			if rs.GetHardState().GetCommit() < os.GetHardState().GetCommit() {
-				return true
-			} else if rs.GetHardState().GetCommit() == os.GetHardState().GetCommit() {
-				// better use voter rather than learner
-				for _, peer := range r.Region().GetPeers() {
-					if peer.StoreId == r.storeID {
-						if peer.Role == metapb.PeerRole_DemotingVoter || peer.Role == metapb.PeerRole_Learner {
-							return true
-						}
+func (r *regionItem) IsRaftStale(origin *regionItem, u *unsafeRecoveryController) bool {
+	cmps := []func(a, b *regionItem) int{
```

Comment:
Why not directly return bool?

Reply:
it continue the cmp range only when the return value is zero, if it's not zero it return v < 0